### PR TITLE
Fix for #3067

### LINKED ISSUE #3067: Internal credentials object prevents setting of a custom own auth header in http request node
### What are the steps to reproduce?

1. Create a HTTP request with a function node as input
2. Set a Authorization header as input message

Example:
   
```javascript
return {
  'method': 'GET',
  'url': 'https://example.com',
  'headers': {
    'Authorization': 'bearer abc123'
  }
};
```
Problem is that the check if credentials are available is always true.

file: packages/node_modules/@node-red/nodes/core/network/21-httprequest.js line 208

```javascript
if (this.credentials) {  // <-- always true -> empty object
```

### What happens?

The Authorization header value will be overwritten with ''Basic " in the http request node.

### What do you expect to happen?

The custom Authorization should be used.

### Please tell us about your environment:

- [X] Node-RED version: 2.0.0
- [X] Node.js version: v16.5.0
- [X] npm version: 7.19.1
- [X] Platform/OS: Arch Linux
- [X] Browser: Firefox

### PR DESCRIPTION
**Pull request by hardillb**
Check if there are any content to the credential object

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Proposed changes

Add to test to see if the http-request's credentials object is empty

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [ ] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

**Comment by knolleary** [comment]
There's some logic on line 212 that pulls username/password out of the URL of they are there, rather than from the credentials object. I think this change will cause that logic to be bypassed. Maybe pull the URL parse out of this if block and use that logic to set `credentials.username/password` if they aren't already set. Then when it gets to the if statement, the props will now exist and they can be applied.

File: packages/node_modules/@node-red/nodes/core/network/21-httprequest.js [hunk 1]
```diff
@@ -205,23 +205,25 @@ module.exports = function(RED) {
                     }
                 }
             }
-            if (this.credentials) {
+            var parsedURL = new URL(url)
+            if (parsedURL.username) {
```
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:209]:
```suggestion
            if (parsedURL.username && !this.credentials.user) {
```

File: packages/node_modules/@node-red/nodes/core/network/21-httprequest.js [hunk 2]
```diff
@@ -205,23 +205,25 @@ module.exports = function(RED) {
...
+                this.credentials.user = parsedURL.username
+            }
+            if (parsedURL.password) {
```
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:212]:
```suggestion
            if (parsedURL.password && !this.credentials.password) {
```

File: packages/node_modules/@node-red/nodes/core/network/21-httprequest.js [hunk 3]
```diff
@@ -205,23 +205,25 @@ module.exports = function(RED) {
                     }
                 }
             }
-            if (this.credentials) {
+            var parsedURL = new URL(url)
```
Comment (knolleary) [inline packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:208]:
```suggestion
            this.credentials = this.credentials || {}
            var parsedURL = new URL(url)
```

**Review by knolleary — CHANGES_REQUESTED** [review]
I think a small change is needed to preserve the existing precedence of `this.credentials` over the url provided details.

File: packages/node_modules/@node-red/nodes/core/network/21-httprequest.js (hunk 3 above)
Comment (hardillb) [reply packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:208]: Not sure we need this one, but better safe than sorry

Comment (hardillb) [reply packages/node_modules/@node-red/nodes/core/network/21-httprequest.js:208]: But adding it does explain the earlier test failure, (I think when run for real `this.credentials` is always `{}` but it's just not added in the tests)
